New wav file selection randomization code

FXF48-Mila opened the pull request.

FXF48-Mila commented:
@leewujung Hi Wu-Jung, I completed the code for the random selection of rows in Bat Detection 2022, taking into account extreme weather conditions (I used the weather dataset from the Mollie's project). Can you view my code? Thank you so much.

leewujung commented:
@FXF48-Mila : Awesome!! Thanks for submitting this! I'll try to get to this as soon as I can, likely during the weekend.

leewujung commented on data/randomization/randomization-with-weather.ipynb:
```diff
@@ -0,0 +1,732 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #6.    dfw = dfw.loc[:, ['Datetime UTC', 'Rain', 'Speed']]
</pre><p>Just nitpicking here: I think you could change "Speed" to "Wind speed" to remove ambiguity when reading this notebook later.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/uw-echospace/bat-detector-validation/pull/6/discussion/#comment-1377770736'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

leewujung replied:
<p>So a good thing to do here (and maybe you already did it to select these criteria?) is to plot the distribution (histogram) of the rain amount and wind speed. Could you add a cell before getting the cell that computes the mean? This way we can easily see what the distributions look like and determine what is a good threshold for the "tail"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/uw-echospace/bat-detector-validation/pull/6/discussion/#comment-1377770740'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

leewujung replied:
<pre class="ql-syntax" spellcheck="false">Line #16.        if len(normal_weather_data) &gt;=2:
</pre><p>Seems that this is a check to make sure there are enough file from a particular site and month? Maybe add a comment about this to make sure it is clear.</p><p> </p><p>It is probably a good idea for you to be able to easily see which site-month combinations have or not have sampled files, as a separate dataframe.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/uw-echospace/bat-detector-validation/pull/6/discussion/#comment-1377770747'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

leewujung replied:
<pre class="ql-syntax" spellcheck="false">Line #20.                selected_rows = selected_rows.append(selected_extreme_row).append(selected_normal_row)
</pre><p>I think it would be good for <code>check_extreme_conditions</code> to return the type of extreme condition(s) the given row contains, since it is heavy rain OR high wind, and put this information into the <code>selected_rows</code> dataframe so that it is easy to see under what condition a specific file is selected (normal or heavy rain and/or high wind)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/uw-echospace/bat-detector-validation/pull/6/discussion/#comment-1377770749'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

leewujung reviewed (commented):
@FXF48-Mila: I like how clean and well-documented your notebook is! Great job!!

I put some inline comments on the notebook using ReviewNB -- the display would make more sense if you look there.

On the output dataframe: I noticed that some files are marked as "Not usable; no metadata" -- could you check what this is, and potentially removing them? Also pinging @aditya-uw here since missing metadata only happens under specific conditions?

aditya-uw commented:
I knew I was pinged but I forgot where I was pinged haha. I wanted to let you know "Not usable; no metadata" is given to empty files (when file size is 0)!

FXF48-Mila commented:
I got it. Let me filter out those empty file rows and then apply a random selection on the rows with vaild files.

FXF48-Mila commented:
Hi @leewujung, I modified my code and added some comments to make it more understandable. I used Mollie's code (under her repo) to re-fetch the up-to-date data for the weather (since I found that her dataset only contains weather data up to October 31, 2022). I also added two histograms and applied quantile to select extreme weather conditions. In the big if/else random selection code, I added some code to return the condition for the selected rows (the conditions are "extreme rain", "extreme wind", and "normal"). I also generated another dataset to show whether a site-month combination has 2 sampling files, and whether it has a file for the extreme weather. I hope my random selection code looks better this time around. Can you view my code again? I appreciate your help a lot.

leewujung commented:
Hey @FXF48-Mila : Thanks for updates! I have 2 comments below:

- it seems that in the section below it is not an if-else situation, since `check_extreme_conditions` checks for an "or" condition, so there is no guarantee that when the condition is "Extreme Wind" it is not also "Extreme Rain" -- as in these conditions could co-exist.
   ```python
  # add a column "Condition" to mark the condition of the selected file
  if selected_extreme_row['Mean Wind Speed'].values[0] >= 11.29:
      selected_rows = selected_rows.append(selected_extreme_row.assign(Condition='Extreme Wind'))
  else:
      selected_rows = selected_rows.append(selected_extreme_row.assign(Condition='Extreme Rain'))
   ```

- since only Telephone Field in July did not have extreme weather condition (based on the last dataframe you showed at the end of the notebook), but all other sites in the same month had extreme weather condition, could you just double check that maybe Telephone Field was not recorded for those days with extreme weather condition?

FXF48-Mila commented:
Hi @leewujung , I fixed the if/else statement, now it should also consider the co-existing extreme weather condition. Also, I found that the reason that Telephone Field in July did not have extreme weather condition is because in July, the audiomoth in Telephone Field recorded only four days, and those four days had no extreme weather during the time of recording.

leewujung commented:
Sounds good! Thanks for looking into it. Feel free to merge this PR -- good job!
